remove jargon and address reviewer critiques (#198)
jaybee84/ml-in-rd: content/06.model-complexity.md

> ### Augmenting machine learning models with statistical techniques to reduce misinterpretation when applied to rare disease

Reviewer: ```suggestion
### Augmenting machine learning models with statistical techniques to reduce misinterpretation
```

Machine learning methods are generally accompanied by a few critical assumptions. 

> First, ML methods assume that the dataset contains equal number of samples for all classes in a dataset (i.e. no imbalance among classes). 

Reviewer: ```suggestion
First, ML methods often work best on datasets that contain equal number of samples for all categories (no "class imbalance").
```

> Second, the given dataset is complete with adequate samples for all features (lack of sparsity).

Reviewer: ```suggestion
Second, the dataset is complete; all samples have measurements for all variables in the dataset (i.e., the dataset is not "sparse", meaning that it is not missing data for some samples).
```

> Third, there is no ambiguity about the labels for the samples in the dataset (i.e. no label-noise, applicable for supervised learning methods).

Reviewer: ```suggestion
Third, there is no ambiguity about the labels for the samples in the dataset (no "label-noise", meaning that, for example, we are confident in the disease of each patient in a dataset).  
```

> Third, there is no ambiguity about the labels for the samples in the dataset (no "label-noise", meaning that, for example, we are confident in the disease of each patient in a dataset).  

Reviewer 2: Do we need this much explanation here in light of what I quoted from the intro in https://github.com/jaybee84/ml-in-rd/pull/198/files#r800215869?

Rare disease datasets, however, violate many of these assumptions.

> There is generally a high class imbalance due to small number of samples for specific classes (e.g., only a few patients with a particular rare disease in a health records dataset), abundant label-noise due to incomplete understanding of the disease, and in many cases sparsity in data.

Reviewer: ```suggestion
There is generally a high _class imbalance_ due to small number of samples for specific classes (e.g., only a few patients with a particular rare disease in a health records dataset), the data are often _sparse,_ and there may be abundant _label-noise_ due to incomplete understanding of the disease.
```

Reviewer: I tried to call out these terms in a way that mirrors the intro to this section.

Thus, applying ML to rare disease data without any accommodations for the above shortcomings may lead to models that lack stability required for reproducibility or simplicity required for interpretation.
In this section, we highlight a few ML techniques that can help manage the challenges in applying ML models applied to rare disease data.

Accommodating class imbalance in datasets and increasing the stability of ML predictions can be achieved through decision tree-based ensemble learning methods (Figure[@fig:2]A-B).
Random forests use resampling (with replacement) based techniques to form a consensus about the important predictive features identified by the decision trees[@https://doi.org/10.1023/A:1010933404324; @doi:10.1186/1472-6947-13-134; @doi:10.1214/aos/1031689014; @doi:10.1177/2045894019890549; @doi:10/btzfh6]. 
Additional approaches like combining random forests with resampling without replacement can generate confidence intervals for the model predictions by iteratively exposing the models to incomplete datasets, mimicking real world cases where most rare disease datasets are incomplete [@doi:10.3390/genes11020226].

> It should be noted here that resampling approaches are most helpful in constructing confidence intervals for algorithms that generate the same outcome everytime they are run (i.e. deterministic models).

Reviewer: ```suggestion
Resampling approaches are most helpful in constructing confidence intervals for algorithms that generate the same outcome every time they are run (i.e. deterministic models).
```

> Resampling approaches are most helpful in constructing confidence intervals for algorithms that generate the same outcome every time they are run (i.e. deterministic models).

Reviewer 2: ```suggestion
Resampling approaches are most helpful in constructing confidence intervals for algorithms that generate the same outcome every time they are run (i.e., deterministic models).
```

For non-deterministic models like decision trees that choose features at random for selecting a path to the outcome, resampling approaches can be helpful in determining robustness or stability of the model. 

> Recent studies suggest that there are limitations to decision tree-based ensemble methods when applied to rare disease datasets, leading to adoption of cascade learning, a variant of ensemble learning.[@doi:10.1007/s11634-019-00354-x; @pmid:30815073]

Reviewer: ```suggestion
Recent studies suggest that there are limitations to decision tree-based ensemble methods when applied to rare disease datasets, leading to adoption of cascade learning, a variant of ensemble learning [@doi:10.1007/s11634-019-00354-x; @pmid:30815073].
```

> For non-deterministic models like decision trees that choose features at random for selecting a path to the outcome, resampling approaches can be helpful in determining robustness or stability of the model. 

Reviewer 2: ```suggestion
For decision trees that choose features at random for selecting a path to the outcome (i.e., are non-deterministic), resampling approaches can be helpful in determining robustness or stability of the model. 
```

In cascade learning, multiple methods leveraging distinct underlying assumptions are used in tandem to capture stable patterns existing in the dataset [@doi:10.1109/cvpr.2001.990537; @doi:10.1007/978-3-540-75175-5_16; @doi:10.1109/icpr.2004.1334680]. 
For example, a cascade learning approach for identifying rare disease patients from electronic health record data utilized independent steps for feature extraction (word2vec [@arxiv:1301.3781]), preliminary prediction with ensembled decision trees, and prediction refinement using data similarity metrics [@pmid:30815073]. 

> Combining these three methods resulted in better performance than other methods when implemented on a silver standard dataset in isolation.

Reviewer: I think this is now the first mention of "silver standard" - it might be confusing?

Author: ```suggestion
Combining these three methods resulted in better performance than other methods in identifying rare patients in the given EHR dataset.
```
```suggestion
Combining these three methods resulted in better performance than other methods when implemented on a silver standard dataset in isolation.
```

Reviewer 2: Here's what the introduction currently says:

> Though researchers strive to train useful and informative models, there are challenges inherent to applying ML to rare disease datasets. For example, training supervised models requires datasets where the phenotype labels have very little uncertainty (or “label-noise”) [doi:10.1093/jamia/ocw028] – termed "gold standard" datasets – but rare disease datasets often come with significant label-noise (e.g., silver standard datasets) due to limits in the current understanding of underlying biology and evolving clinical definitions of many rare diseases. Label-noise can decrease prediction accuracy and require larger samples sizes during training [doi:10.1109/tnnls.2013.2292894].

Author: Thanks Jackie.. I thought we had explained it before but had not looked back to confirm :(

> Other approaches like inverse class weighting, oversampling (with methods like Synthetic Minority Oversampling Technique, abbreviated SMOTE), or methods that adequately represent the rarer classes are also beneficial in increasing stability of predictions while accommodating for class imbalance.

Reviewer: As a new-to-this person, I don't think I'd know what inverse class weighting or oversampling mean. Should we bother mentioning these if we don't explain them, or if we don't have any examples of them being used in rare disease?

Or maybe just reordering the sentence to make it gentler: 

```suggestion
Other approaches that better represent rare classes like inverse class weighting, oversampling [doi:10.1613/jair.953], or cascade learning may also improve ML models that use rare disease data. 
```

"methods that adequately represent the rarer classes " I put cascade learning back in because this describes all of these methods, not just cascade learning.

Reviewer: I also removed SMOTE - we can just cite it since we don't talk about it elsewhere AFAIK

Reviewer 2: Agree with Robert's suggestions

> Presence of label-noise and sparsity in the data can also lead to overfitting of models to the training data, such that the models show high prediction accuracy on training data but low prediction accuracy on encountering new data points. 

Reviewer: ```suggestion
The presence of label-noise and sparsity in the data can also lead to overfitting of models to the training data, meaning that the models show high prediction accuracy on the training data but low prediction accuracy on new data. 
```

> Features identified as predictive by overfit models are often not representative of the data making the model not generalizable to new data. [@doi:10.1073/pnas.1900654116]

Reviewer: ```suggestion
Overfit models rely on variables that are unique to the training data (for example, the calibration of the instrument that was used to generate the training data); when models are overfit, they are not generalizable to new data (e.g., data generated on the same instrument that has been recalibrated). [@doi:10.1073/pnas.1900654116]
```

Reviewer: I thought an example would help here that our potential audience could easily understand. Feel free to change the example.

> In such cases, regularization can not only protect ML models from poor generalizability caused by overfitting, but also reduce model complexity by reducing the feature space available for training. (Figure[@fig:2]C)

Reviewer: Do we define regularization before this?

Author: We define regularization in line 32 right after this. Let me know if that seems insufficient.. I tried a couple of different ways to define it but in case we want to include more details about regularization, we would need to go into details of shrinkage of coefficient estimates. IMO that would add more jargon than desirable.

> In such cases, regularization can not only protect ML models from poor generalizability caused by overfitting, but also reduce model complexity by reducing the feature space available for training. (Figure[@fig:2]C)

Author: ```suggestion
In such cases, regularization can not only protect ML models from poor generalizability caused by overfitting, but also reduce model complexity by reducing the feature space available for training. (Figure[@fig:2]C)
Overfit models generate large errors when the model tries to predict values for new evaluation data.
Regularization is a collection of methods that can help reduce these prediction errors by various approaches. 
Examples of some regularization methods include ridge regression, LASSO regression, and Elastic-net regression.
```

> Regularization is often used in rare variant discovery and immune cell signature discovery studies which are useful examples that also need to accomodate sparsity of data like rare diseases.

Reviewer: ```suggestion
Regularization is often used in rare variant discovery and immune cell signature discovery studies which are useful examples that also need to accommodate sparsity of data like rare diseases.
```

> For example, among various regularization methods (ridge regression, LASSO regression, and elastic-net)[@doi:10.1111/j.1467-9868.2005.00503.x], hybrid applications of LASSO for capturing combinations of rare variants have proven beneficial [@doi:10.1016/j.ajhg.2008.06.024; @doi:10.1186/1753-6561-5-s9-s113] .

Reviewer: This section gets jargon-y real fast but is also fairly repetitive across these four sentences. I think this could use an intro sentence describing what regularization is, listing these examples. Then a sentence describing selected applications of these methods in rare disease. Then a sentence saying - you should consider these methods too (basically the last sentence, but it could be shorter).

### Augmenting machine learning models with statistical techniques to reduce misinterpretation

Machine learning methods are generally accompanied by a few critical assumptions. 
First, ML methods often work best on datasets that contain equal number of samples for all categories (no "class imbalance").
Second, the dataset is complete; all samples have measurements for all variables in the dataset (i.e., the dataset is not "sparse", meaning that it is not missing data for some samples).

> Third, there is no ambiguity about the labels for the samples in the dataset (no "label-noise", meaning that, for example, we are confident in the disease of each patient in a dataset).  

Author: ```suggestion
Third, there is no ambiguity about the labels for the samples in the dataset (i.e. no "label-noise").  
```

> ### Augmenting machine learning models with statistical techniques to reduce misinterpretation

Reviewer 2: ```suggestion
### Minimize misinterpretation with statistical techniques
```

Minimize seems too strong a word, perhaps...

Author: ```suggestion
### Reducing misinterpretation of model output with statistical techniques 
```

Recent studies suggest that there are limitations to decision tree-based ensemble methods when applied to rare disease datasets, leading to adoption of cascade learning, a variant of ensemble learning [@doi:10.1007/s11634-019-00354-x; @pmid:30815073].
In cascade learning, multiple methods leveraging distinct underlying assumptions are used in tandem to capture stable patterns existing in the dataset [@doi:10.1109/cvpr.2001.990537; @doi:10.1007/978-3-540-75175-5_16; @doi:10.1109/icpr.2004.1334680]. 
For example, a cascade learning approach for identifying rare disease patients from electronic health record data utilized independent steps for feature extraction (word2vec [@arxiv:1301.3781]), followed by preliminary prediction with ensembled decision trees, and then prediction refinement using data similarity metrics [@pmid:30815073]. 
Combining these three unrelated methods resulted in better overall prediction when implemented on a silver standard dataset, than other methods applied in isolation.

> In addition to cascade learning, other approaches that better represent rare classes like inverse class weighting, oversampling [doi:10.1613/jair.953] may also improve ML models that use rare disease data. 

Reviewer 2: ```suggestion
In addition to cascade learning, other approaches that better represent rare classes like inverse class weighting and oversampling [doi:10.1613/jair.953] may also improve ML models that use rare disease data. 
```

The presence of label-noise and sparsity in the data can also lead to overfitting of models to the training data, meaning that the models show high prediction accuracy on the training data but low prediction accuracy (and large prediction errors) on new evaluation data. 
Overfit models tend to rely on variables that are unique to the training data (for example, the calibration of the instrument that was used to generate the training data), and not generalizable to new data (e.g., data generated on the same instrument that has been recalibrated). [@doi:10.1073/pnas.1900654116]
In such cases, regularization can not only protect ML models from poor generalizability caused by overfitting, but also reduce model complexity by reducing the feature space available for training. (Figure[@fig:2]C)

> Regularization is a collection of methods (generally regression methods) that can help reduce prediction errors by various approaches. 

Reviewer 2: I don't know that I agree that "regularization is a collection of methods (generally regression methods)" vs. it's the approach/concept of adding a penalty to increase generalizability or some other desirable property and control overfitting. So if we take a VAE as an example, use of KL divergence in the loss function is often considered/talked about as regularization.

Reviewer 2: I think this comment could be addressed by updating this sentence to be more general. I think using the regression examples is fine. But then maybe 1-2 sentences that talk about other scenarios at the end of the paragraph...

Author: Fair point..

Author: ```suggestion
Regularization is an approach by which a penalty is added to the model to avoid making large prediction errors. 
```

Author: @jaclyn-taroni do you have an example where KL divergence or other types of regularizations may have been used in the context of rare disease or rare dataset? I would like to add a sentence towards the end of this section to prompt readers to explore regularization methods other than L1, L2 and elastic net.. but most of the rare disease/dataset papers I came across had a variant of the above mentioned regressions.

> Examples of some regularization approaches include ridge regression, LASSO regression, and Elastic-net regression.

Author: ```suggestion
Some examples of some regularization approaches include ridge regression, LASSO regression, and Elastic-net regression, among others.
```

Regularization is often used in rare variant discovery and immune cell signature discovery studies which are useful examples that also need to accommodate sparsity of data like rare diseases.
For example, among various regularization methods hybrid applications of LASSO for capturing combinations of rare and common variants associated with specific traits have proven beneficial [@doi:10.1186/1753-6561-5-s9-s113].
In this example, applying LASSO regularization reduced the number of common variants included as features in the final analysis generating a simpler model while reducing error in the association of common and rare variants with a specific trait. 
In the context of rare immune cell signature discovery, variations of elastic-net regression were found to outperform other regression approaches [@doi:10.1111/j.1467-9868.2005.00503.x; @doi:10.1016/j.compbiomed.2015.10.008; @doi:10.1186/s12859-019-2994-z]. 

> Thus regularization methods like LASSO or elastic-net have been methods of choice while working with rare observations.

Author: ```suggestion
Thus regularization methods like LASSO or elastic-net have been methods of choice while working with rare observations.[@doi:10.1371/journal.pgen.1004754]
```
